common/ledger/snapshot/file.go

```diff
@@ -0,0 +1,200 @@
+	}
+	return &FileCreator{
+		file:              file,
+		bufWriter:         bufWriter,
+		multiWriter:       multiWriter,
+		hasher:            hasher,
+		varintReusableBuf: make([]byte, binary.MaxVarintLen64),
+	}, nil
+}
+
+// EncodeString encodes and appends the string to the data stream
+func (c *FileCreator) EncodeString(str string) error {
+	return c.EncodeBytes([]byte(str))
+}
+
+// EncodeString encodes and appends a proto message to the data stream
+func (c *FileCreator) EncodeProtoMessage(m proto.Message) error {
+	b, err := proto.Marshal(m)
+	if err != nil {
+		return errors.Wrapf(err, "error marshalling proto message to write to the snapshot file: %s", c.file.Name())
+	}
+	return c.EncodeBytes(b)
+}
+
+// EncodeBytes encodes and appends bytes to the data stream
+func (c *FileCreator) EncodeBytes(b []byte) error {
+	if err := c.EncodeUVarint(uint64(len(b))); err != nil {
+		return err
+	}
+	if _, err := c.multiWriter.Write(b); err != nil {
+		return errors.Wrapf(err, "error while writing data to the snapshot file: %s", c.file.Name())
+	}
+	return nil
+}
+
+// EncodeUVarint encodes and appends a number to the data stream
+func (c *FileCreator) EncodeUVarint(u uint64) error {
+	n := binary.PutUvarint(c.varintReusableBuf, u)
+	if _, err := c.multiWriter.Write(c.varintReusableBuf[:n]); err != nil {
+		return errors.Wrapf(err, "error while writing data to the snapshot file: %s", c.file.Name())
+	}
+	return nil
+}
+
+// Done closes the snapshot file and returns the final hash of the data stream
+func (c *FileCreator) Done() ([]byte, error) {
+	if err := c.bufWriter.Flush(); err != nil {
+		return nil, errors.Wrapf(err, "error while flushing to the snapshot file: %s ", c.file.Name())
+	}
+	if err := c.file.Close(); err != nil {
+		return nil, errors.Wrapf(err, "error while closing the snapshot file: %s ", c.file.Name())
+	}
+	return c.hasher.Sum(nil), nil
+}
+
+// Close closes the underlying file, if not already done. A consumer can invoke this function if the consumer
+// encountered some error and simply wants to abandon the snapshot file creation (typically, intended to be used in a defer statement)
+func (c *FileCreator) Close() error {
+	if c == nil {
+		return nil
+	}
+	return errors.Wrapf(c.file.Close(), "error while closing the snapshot file: %s", c.file.Name())
+}
+
+// FileReader reads from a ledger snapshot file. This is expected to be used for loading the ledger snapshot data
+// during bootstrapping a channel from snapshot. The data should be read, using the functions `DecodeXXX`,
+// in the same sequence in which the data was written by the functions `EncodeXXX` in the `FileCreator`.
+// Note that the FileReader does not verifies the hash of stream and it is expected that the hash has been verified
+// by the consumer. Later, if we decide to perform this, on-the-side, while loading the snapshot data, the FileRedear,
+// like the FileCreator, would take a `hasher` as an input
+type FileReader struct {
```

Comment:
I just realized that the batching is not that easy as we prepend length to each item. If we need batching during reads, we need to do the same during the write. 

Otherwise, we can also read a few MB into a buffer and read one by one from it.

Reply:
Yes, that's what `bufReader` is does here under the cover.

Reply:
I see. The `defaultBufSize` is 4096 bytes (i.e., 1 disk block). That is applicable for even a single byte read from a file as OS driver reads per disk block. Are you saying that if we increase the `defaultBufSize` using `NewReaderSize()`, it would read more blocks?

Reply:
Yes, that's how it works.